### PR TITLE
Return if column is generated or not

### PR DESCRIPTION
**Pull request by devotis**
Related to
https://github.com/strongloop/loopback-datasource-juggler/issues/899

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by superkhau** [comment]
Can you add some tests to verify your changes and prevent regressions in the future?

**Comment by devotis** [comment]
@superkhau I've added tests

File: test/mysql.discover.test.js [hunk 1]
```diff
@@ -206,6 +206,42 @@ describe('Discover model foreign keys', function() {
   });
 });
 
+describe('Discover model generated columns', function() {
+  it('should return an array of columns for STRONGLOOP.PRODUCT and none of them is generated', function(done) {
```
Comment (superkhau) [inline test/mysql.discover.test.js]: it('returns an array of columns when the table has no generated columns'

File: test/mysql.discover.test.js [hunk 2]
```diff
@@ -206,6 +206,42 @@ describe('Discover model foreign keys', function() {
...
+    db.discoverModelProperties('product', function(err, models) {
+      if (err) {
+        console.error(err);
+        done(err);
```
Comment (superkhau) [inline test/mysql.discover.test.js]: if (err) return done(err);

File: test/mysql.discover.test.js [hunk 3]
```diff
@@ -206,6 +206,42 @@ describe('Discover model foreign keys', function() {
...
+      } else {
```
Comment (superkhau) [inline test/mysql.discover.test.js]: remove else

File: test/mysql.discover.test.js [hunk 4]
```diff
@@ -206,6 +206,42 @@ describe('Discover model foreign keys', function() {
...
+        models.forEach(function(m) {
```
Comment (superkhau) [inline test/mysql.discover.test.js]: m -> model

File: test/mysql.discover.test.js [hunk 5]
```diff
@@ -206,6 +206,42 @@ describe('Discover model foreign keys', function() {
...
+          // console.dir(m);
```
Comment (superkhau) [inline test/mysql.discover.test.js]: remove comment

File: test/mysql.discover.test.js [hunk 6]
```diff
@@ -206,6 +206,42 @@ describe('Discover model foreign keys', function() {
...
+          assert(m.tableName === 'PRODUCT');
+          assert(!m.generated, 'STRONGLOOP.PRODUCT table should not have generated (identity) columns');
+
+        });
+        done(null, models);
```
Comment (superkhau) [inline test/mysql.discover.test.js]: done();

File: test/mysql.discover.test.js [hunk 7]
```diff
@@ -206,6 +206,42 @@ describe('Discover model foreign keys', function() {
...
+      }
+    });
+  });
+  it('should return an array of columns for STRONGLOOP.TESTGEN and the first is generated', function(done) {
```
Comment (superkhau) [inline test/mysql.discover.test.js]: it('returns an array of columns when the first column of the table is generated'

File: test/mysql.discover.test.js [hunk 8]
```diff
@@ -206,6 +206,42 @@ describe('Discover model foreign keys', function() {
...
+    db.discoverModelProperties('testgen', function(err, models) {
+      if (err) {
+        console.error(err);
+        done(err);
```
Comment (superkhau) [inline test/mysql.discover.test.js]: if (err) return done(err);

File: test/mysql.discover.test.js [hunk 9]
```diff
@@ -206,6 +206,42 @@ describe('Discover model foreign keys', function() {
...
+      } else {
```
Comment (superkhau) [inline test/mysql.discover.test.js]: remove else

File: test/mysql.discover.test.js [hunk 10]
```diff
@@ -206,6 +206,42 @@ describe('Discover model foreign keys', function() {
...
+        models.forEach(function(m) {
```
Comment (superkhau) [inline test/mysql.discover.test.js]: m -> model

File: test/mysql.discover.test.js [hunk 11]
```diff
@@ -206,6 +206,42 @@ describe('Discover model foreign keys', function() {
...
+          // console.dir(m);
+          assert(m.tableName === 'TESTGEN');
+          if (m.columnName === 'ID') {
+            assert(m.generated, 'STRONGLOOP.TESTGEN.ID should be a generated (identity) column');
+          }
+        });
+        done(null, models);
```
Comment (superkhau) [inline test/mysql.discover.test.js]: done();

**Comment by devotis** [comment]
Just rebased. Will now implement your review comments...

File: test/mysql.discover.test.js (hunk 1 above)
Comment (devotis) [inline test/mysql.discover.test.js]: Actually what your saying is incorrect. An array of columns is returned regardless. I'm testing here that none of the columns in the array have generated=true. The description of the test should express that. I could omit STRONGLOOP.PRODUCT in the description if you find that important, but as you can see all the other tests are expressed just the same.

**Comment by devotis** [comment]
@superkhau Basically all your comments instruct me to diverge from the code conventions in the rest of the file. I mean the code convention that I used is exactly the same as al the other tests in the same file. You really want met to create inconsistency in the code convention used in a single file?

**Comment by superkhau** [comment]
@devotis You're right about the consistency within a single file -- we can leave it for now -- we have to clean this up eventually to match the conventions I mentioned above at some point but that is outside the scope of this PR. ;) I approve your changes. :+1:

**Comment by superkhau** [comment]
@slnode test please

**Comment by loay** [comment]
@slnode test please

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by superkhau** [comment]
@kjdelisle Can you help land this at some point? I'm reassigning to your team.

**Comment by b-admike** [comment]
@slnode test please

**Comment by b-admike** [comment]
@slnode test please

**Comment by kjdelisle** [comment]
@slnode test please

**Comment by b-admike** [comment]
@slnode test please

File: lib/discovery.js [hunk 12]
```diff
@@ -204,7 +204,8 @@ function mixinDiscovery(MySQL, mysql) {
         ' numeric_precision AS "dataPrecision",' +
         ' numeric_scale AS "dataScale",' +
         ' column_type AS "columnType",' +
-        ' is_nullable = \'YES\' AS "nullable"' +
+        ' is_nullable = \'YES\' AS "nullable",' +
+        ' CASE WHEN extra LIKE \'%auto_increment%\' THEN 1 ELSE 0 END AS generated' +
```
Comment (loay) [inline lib/discovery.js]: "generated"

File: lib/discovery.js [hunk 13]
```diff
@@ -218,7 +219,8 @@ function mixinDiscovery(MySQL, mysql) {
...
```
Comment (loay) [inline lib/discovery.js]: "generated"

**Comment by loay** [comment]
Hi @devotis  
Please address the changes that I requested .. it is just a minor syntax error. 
then rebase and that should fix the test failures. Thanks.

**Comment by devotis** [comment]
Just fixed it. Do you know if the 'Update branch' button rebases?

image

**Comment by devotis** [comment]
@loay I'm done rebasing...

**Comment by loay** [comment]
@slnode test please

File: test/mysql.discover.test.js [hunk 14]
```diff
@@ -206,6 +206,42 @@ describe('Discover model foreign keys', function() {
...
+    db.discoverModelProperties('product', function(err, models) {
+      if (err) {
+        console.error(err);
+        done(err);
+      } else {
+        models.forEach(function(m) {
+          // console.dir(m);
+          assert(m.tableName === 'PRODUCT');
+          assert(!m.generated, 'STRONGLOOP.PRODUCT table should not have generated (identity) columns');
+
```
Comment (loay) [inline test/mysql.discover.test.js]: remove that lone please

**Comment by loay** [comment]
tests pass on one server but there is a linting problem:

/test/mysql.discover.test.js
  221:9  error  Block must not be padded by blank lines  padded-blocks

You can run this command `./node_modules/.bin/eslint --fix .` in your terminal before committing and it should fix your linting issues. I pointed out where the issue is anyways.

**Comment by devotis** [comment]
I don't even know if this works, but I'm gonna ask anyway:

@slnode test please

:)

**Comment by loay** [comment]
@slnode test please

**Comment by loay** [comment]
@devotis Almost there ..
More tests are passing now, but we still have 2 failures on the windows server that we need to figure out.

**Comment by loay** [comment]
@slnode test please

**Comment by ssh24** [comment]
@devotis Could you please rebase your branch?

**Comment by ssh24** [comment]
@devotis Could you please rebase your branch and fix the commit linter errors?

```
**  1 problems found:
**    c2057ac - Fix 'Block must not be padded by blank lines padded: First line should be 50 characters or less (saw 59)
```

**Comment by ssh24** [comment]
Hi @devotis,

Could you follow https://github.com/blog/2247-improving-collaboration-with-forks and give me access to the forked repo, then I can handle rebasing for you?

If this cannot be done, I am sorry but this PR needs to be closed as it is taking away too long to land it.

**Comment by devotis** [comment]
I was on holiday. Still on mobile. I'll check back in from my computer
tomorrow or the day after.

**Comment by devotis** [comment]
@ssh24 I've added you as a collaborator to my fork. Can you handle the rebase and commit a fix for the linting error?

**Comment by ssh24** [comment]
@slnode ok to test

**Comment by ssh24** [comment]
@devotis Thank you. I have rebased your branch. The tests that you have added along with your changes are failing on CI. Could you please take a look at those tests and patch a fix for it?

**Comment by ssh24** [comment]
@devotis Could you please fix the commit linter error? In the mean time, I will run the tests on CI and see if it passes.
```
**************************************************
**
**  Linting commit logs
**
**  1 problems found:
**    a6f3b54 - Should fix the tests failing on the tableName (inco: First line should be 50 characters or less (saw 62)
**
**************************************************
```

**Comment by ssh24** [comment]
@slnode ok to test

**Comment by devotis** [comment]
@ssh24 can you do it for me? It's a minor, but I'm not available for this for the next day.

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@devotis PR landed. Thank you for the contribution!

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Pull request merged**